File: services/worker/telemetry.go
package main

import (
	"context"
	"os"
	"strings"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.38.0"
)

func startTelemetry(ctx context.Context) func(context.Context) error {
	endpoint := strings.TrimSuffix(strings.TrimSpace(os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT")), "/")
	if endpoint == "" {
		return func(context.Context) error { return nil }
	}
	exporter, err := otlptracehttp.New(ctx, otlptracehttp.WithEndpointURL(endpoint+"/v1/traces"))
	if err != nil {
		return func(context.Context) error { return nil }
	}
	provider := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithResource(resource.NewWithAttributes("", semconv.ServiceName("durable-agent-runtime-worker"))),
	)
	otel.SetTracerProvider(provider)
	return provider.Shutdown
}

File: services/control-plane/src/telemetry.ts
import { trace } from "@opentelemetry/api";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import { resourceFromAttributes } from "@opentelemetry/resources";
import {
  BatchSpanProcessor,
  NodeTracerProvider,
} from "@opentelemetry/sdk-trace-node";

let shutdown: (() => Promise<void>) | undefined;

export function startTelemetry(): void {
  const endpoint = process.env.OTEL_EXPORTER_OTLP_ENDPOINT;
  if (!endpoint || shutdown) return;
  const provider = new NodeTracerProvider({
    resource: resourceFromAttributes({
      "service.name": "durable-agent-runtime-control-plane",
    }),
    spanProcessors: [
      new BatchSpanProcessor(
        new OTLPTraceExporter({
          url: `${endpoint.replace(/\/$/, "")}/v1/traces`,
        }),
      ),
    ],
  });
  provider.register();
  shutdown = () => provider.shutdown();
}

export function requestTracer() {
  return trace.getTracer("durable-agent-runtime.control-plane");
}

export async function shutdownTelemetry(): Promise<void> {
  await shutdown?.();
}
